onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -0,0 +1,515 @@
+            [&](size_t i) { scale[i] = QuantBScale[i * StrideQuantBScale]; }
+        );
+
+        float offset[NCols];  // Includes zero point and float conversion offset of 16.
+                              // More details on the float conversion below.
+        if (QuantBZeroPointColPtr != nullptr) {
+            UnrolledLoop<NCols>([&](size_t i) {
+                const uint8_t zp_packed =
+                    QuantBZeroPointColPtr[i * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+                const uint8_t zp = ((QuantBZeroPointIdx & 1) == 1) ? (zp_packed >> 4) : (zp_packed & 0x0F);
+                offset[i] = 16.0f + zp;
+            });
+        } else {
+            UnrolledLoop<NCols>([&](size_t i) {
+                constexpr uint8_t zp = 8;
+                offset[i] = 16.0f + zp;
+            });
+        }
+
+        constexpr size_t SubBlkLen = 16;  // number of block elements to process in one iteration
+
+        for (size_t k_idx_in_blk = 0; k_idx_in_blk < k_blk_len; k_idx_in_blk += SubBlkLen) {
+            // load A row vector elements
+
+            // load `SubBlkLen` elements from A padding with 0's if there aren't enough
+            const size_t k_subblk_len = std::min(k_blk_len - k_idx_in_blk, SubBlkLen);
+            const auto av = LoadDataAndZeroPad<SubBlkLen>(ARowPtr + k + k_idx_in_blk, k_subblk_len);
+
+            // load B column vectors
+            uint8x8_t bv_packed[NCols];
+            UnrolledLoop<NCols>([&](size_t i) { bv_packed[i] = vld1_u8(b_data[i]); });
+
+            uint8x8_t bv_u8_unzipped[NCols][2];
+            UnrolledLoop<NCols>([&](size_t i) {
+                bv_u8_unzipped[i][0] = vand_u8(bv_packed[i], LowMask);
+                bv_u8_unzipped[i][1] = vand_u8(vshr_n_u8(bv_packed[i], 4), LowMask);
+            });
+
+            uint8x8_t bv_u8[NCols][2];
+            UnrolledLoop<NCols>([&](size_t i) {
+                bv_u8[i][0] = vzip1_u8(bv_u8_unzipped[i][0], bv_u8_unzipped[i][1]);
+                bv_u8[i][1] = vzip2_u8(bv_u8_unzipped[i][0], bv_u8_unzipped[i][1]);
+            });
+
+            // dequantize B
+
+            // Manual conversion to float takes place in two steps:
+            // 1. Map 4-bit values from [0, 15] to float values from [16.0f, 31.0f].
+            //    This target float range is convenient because the 4-bit source values can be placed directly into the
+            //    target float bits.
+            // 2. Subtract the conversion offset of 16 from the float result.
+
+            // The high 16 bits of an IEEE 754 32-bit float used as a template for creating float values.
+            constexpr uint16_t float_high_half_template = 0b0'10000011'0000000;
+            //                                           sign|exponent|partial mantissa
+            //                                              +|131: 2^4|~~~~ <- 4 bits go here
+
+            const uint16x8_t float_high_half_template_v = vdupq_n_u16(float_high_half_template);
```

Comment:
>const [](http://example.com/codeflow?start=12&length=5)

nit: constexpr

Reply:
unfortunately it doesn't:
`error C2131: expression did not evaluate to a constant`

Reply:
move this to the beginning of the function then

Reply:
sure, moved